bin/propolis-server/src/lib/serial/backend.rs

```diff
@@ -0,0 +1,397 @@
+
+//! A backend that provides external clients with access to a Propolis character
+//! device and tracks the history of bytes the guest has written to that device.
+
+use std::{
+    collections::BTreeMap,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+};
+
+use propolis::chardev::{
+    pollers::{SinkBuffer, SourceBuffer},
+    Sink, Source,
+};
+use tokio::{
+    select,
+    sync::{mpsc, oneshot},
+};
+
+use super::history_buffer::{HistoryBuffer, SerialHistoryOffset};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct ClientId(u64);
+
+/// A client's rights when accessing a character backend.
+#[derive(Clone, Copy)]
+pub(super) enum Permissions {
+    /// The client may both read and write to the backend.
+    ReadWrite,
+
+    /// The client may only read from the backend.
+    ReadOnly,
+}
+
+impl Permissions {
+    pub(super) fn is_writable(&self) -> bool {
+        matches!(self, Permissions::ReadWrite)
+    }
+}
+
+/// Determines what happens when the backend is unable to send a guest byte back
+/// to a client because the client's channel is full.
+#[derive(Clone, Copy)]
+pub(super) enum FullReadChannelDiscipline {
+    /// The backend should block until it can send to this client.
+    Block,
+
+    /// The backend should close this client's connection.
+    Close,
+}
+
+/// An individual client of a character backend.
+struct Client {
+    /// Bytes read from the character device should be sent to the client on
+    /// this channel.
+    tx: mpsc::Sender<u8>,
+
+    /// Determines what happens when the backend wants to send a byte and
+    /// [`Self::tx`] is full.
+    read_discipline: FullReadChannelDiscipline,
+}
+
+/// A handle held by a client that represents its connection to the backend.
+pub(super) struct ClientHandle {
+    /// The client's ID.
+    id: ClientId,
+
+    /// A reference to the backend to which the client is connected.
+    backend: Arc<ConsoleBackend>,
+
+    /// The client's backend access rights.
+    permissions: Permissions,
+}
+
+impl ClientHandle {
+    pub(super) fn is_writable(&self) -> bool {
+        self.permissions.is_writable()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let mut inner = self.backend.inner.lock().unwrap();
+        inner.clients.remove(&self.id);
+    }
+}
+
+impl ClientHandle {
+    /// Attempts to write the bytes in `buf` to the backend.
+    ///
+    /// The backend may buffer some of the written bytes before sending them to
+    /// its device, so when this function returns, it is not guaranteed that all
+    /// of the written bytes have actually reached the guest.
+    ///
+    /// # Return value
+    ///
+    /// - `Ok(bytes)` if the write succeeded; `bytes` is the number of bytes
+    ///   that were written.
+    /// - `Err(ErrorKind::PermissionDenied)` if this handle does not grant write
+    ///   access to the device.
+    /// - `Err(ErrorKind::ConnectionAborted)` if this handle's client was
```

Comment:
nitpicky, considered unimportant: this doc comment is *technically* lying; this doesn't return an `ErrorKind`, but a whole `io::Error`...